post-finding security hub has issues with isoformat on ec2, asg and failing the lambda

**Describe the bug**
The lambda generated to post finding on Securityhub is failing on isoFormat 

`2020-09-30T10:13:04.829-04:00Copy[ERROR]	2020-09-30T14:13:04.828Z	6f0773e7-732f-4784-b4a1-182702d42ba6	error during policy executionTraceback (most recent call last):  File "/var/task/c7n/handler.py", line 166, in dispatch_event    p.push(event, context)  File "/var/task/c7n/policy.py", line 1134, in push    return mode.run(event, lambda_ctx)  File "/var/task/c7n/policy.py", line 850, in run    resources = super(ConfigRuleMode, self).run(event, lambda_context)  File "/var/task/c7n/policy.py", line 450, in run    return self.run_resource_set(event, resources)  File "/var/task/c7n/policy.py", line 480, in run_resource_set    results = action.process(resources)  File "/var/task/c7n/resources/securityhub.py", line 409, in process    finding = self.get_finding(  File "/var/task/c7n/resources/securityhub.py", line 531, in get_finding    finding_resources.append(self.format_resource(r))  File "/var/task/c7n/resources/asg.py", line 739, in format_resource    details['CreatedTime'] = details['CreatedTime'].isoformat()AttributeError: 'int' object has no attribute 'isoformat' | [ERROR] 2020-09-30T14:13:04.828Z 6f0773e7-732f-4784-b4a1-182702d42ba6 error during policy execution Traceback (most recent call last): File "/var/task/c7n/handler.py", line 166, in dispatch_event p.push(event, context) File "/var/task/c7n/policy.py", line 1134, in push return mode.run(event, lambda_ctx) File "/var/task/c7n/policy.py", line 850, in run resources = super(ConfigRuleMode, self).run(event, lambda_context) File "/var/task/c7n/policy.py", line 450, in run return self.run_resource_set(event, resources) File "/var/task/c7n/policy.py", line 480, in run_resource_set results = action.process(resources) File "/var/task/c7n/resources/securityhub.py", line 409, in process finding = self.get_finding( File "/var/task/c7n/resources/securityhub.py", line 531, in get_finding finding_resources.append(self.format_resource(r)) File "/var/task/c7n/resources/asg.py", line 739, in format_resource details['CreatedTime'] = details['CreatedTime'].isoformat() AttributeError: 'int' object has no attribute 'isoformat'`

-- | --




**To Reproduce**
used anchor for post finding
```
 post-finding-failed: &post-finding-failed  
    type: post-finding
    description: |
        mandatory tags are required on S3.
    title: "custodian-s3-mandatory-tags-missing"
    severity_normalized: 70
    types:
      - "Software and Configuration Checks/AWS Security Best Practices"
    recommendation: "Add data confidentiality tag to S3 buckets"
    recommendation_url: "https://wikiurk"
    compliance_status: FAILED
    severity_label: HIGH
```

Policy

```
 - name: ec2-mandatory-tag-check
    resource: ec2
    description: |
       The policy checks the ec2 required tags
    mode: *config-mode
    filters:
      - and: *filters-absent
      - "tag:aws:autoscaling:groupName": absent
    actions:
      - <<: *post-finding-failed
        title: "custodian-ec2-mandatory-tags-missing"
        description: "Mandatory tags are required on ec2"
      # - type: mark-for-op
      #   op: stop
      #   days: 7 
      - <<: *notify
        violation_desc: "Mandatory Tags are missing on EC2"
```

ASG
```
 - name: asg-ec2-mandatory-tag-check
    resource: asg
    description: |
       The policy checks the asg required tags
    mode: *config-mode
    filters:
      - and: *filters-absent
    actions:
      - <<: *post-finding-failed
        title: "custodian-asg-ec2-mandatory-tags-missing"
        description: "Mandatory tags are required on asg"
      # - type: mark-for-op
      #   op: suspend
      #   days: 7 
      - <<: *notify
        violation_desc: "Mandatory Tags are missing on ASG"
```



**Expected behavior**
Lambda should publish to security hub

**Background (please complete the following information):**
 - Python Version: [e.g. python 3.8.1]
 - Custodian Version: [e.g. 0.9.4]
 - Tool Version: [if applicable]
 - Cloud Provider: [e.g. gcp, aws, azure]
 - Policy: [please exclude any account/sensitive information]
```yaml
policies: 
   - name: check-buckets
      resource: aws.s3
```
 - Traceback: [if applicable, please exclude sensitive/account information]
 - `custodian version --debug` output

**Additional context**
Add any other context about the problem here.

I have tried with 0.9.6 and a very simple post-finding ..it seems like its broken


`[ERROR] AttributeError: 'str' object has no attribute 'isoformat'Traceback (most recent call last):  File "/var/task/custodian_policy.py", line 4, in run    return handler.dispatch_event(event, context)  File "/var/task/c7n/handler.py", line 166, in dispatch_event    p.push(event, context)  File "/var/task/c7n/policy.py", line 1134, in push    return mode.run(event, lambda_ctx)  File "/var/task/c7n/policy.py", line 850, in run    resources = super(ConfigRuleMode, self).run(event, lambda_context)  File "/var/task/c7n/policy.py", line 450, in run    return self.run_resource_set(event, resources)  File "/var/task/c7n/policy.py", line 480, in run_resource_set    results = action.process(resources)  File "/var/task/c7n/resources/securityhub.py", line 409, in process    finding = self.get_finding(  File "/var/task/c7n/resources/securityhub.py", line 531, in get_finding    finding_resources.append(self.format_resource(r))  File "/var/task/c7n/resources/ec2.py", line 1145, in format_resource    "LaunchedAt": r["LaunchTime"].isoformat() | [ERROR] AttributeError: 'str' object has no attribute 'isoformat' Traceback (most recent call last):   File "/var/task/custodian_policy.py", line 4, in run     return handler.dispatch_event(event, context)   File "/var/task/c7n/handler.py", line 166, in dispatch_event     p.push(event, context)   File "/var/task/c7n/policy.py", line 1134, in push     return mode.run(event, lambda_ctx)   File "/var/task/c7n/policy.py", line 850, in run     resources = super(ConfigRuleMode, self).run(event, lambda_context)   File "/var/task/c7n/policy.py", line 450, in run     return self.run_resource_set(event, resources)   File "/var/task/c7n/policy.py", line 480, in run_resource_set     results = action.process(resources)   File "/var/task/c7n/resources/securityhub.py", line 409, in process     finding = self.get_finding(   File "/var/task/c7n/resources/securityhub.py", line 531, in get_finding     finding_resources.append(self.format_resource(r))   File "/var/task/c7n/resources/ec2.py", line 1145, in format_resource     "LaunchedAt": r["LaunchTime"].isoformat()`
-- | --

same for EBS
[ERROR] AttributeError: 'str' object has no attribute 'isoformat'
Traceback (most recent call last):
  File "/var/task/custodian_policy.py", line 4, in run
    return handler.dispatch_event(event, context)
  File "/var/task/c7n/handler.py", line 166, in dispatch_event
    p.push(event, context)
  File "/var/task/c7n/policy.py", line 1134, in push
    return mode.run(event, lambda_ctx)
  File "/var/task/c7n/policy.py", line 850, in run
    resources = super(ConfigRuleMode, self).run(event, lambda_context)
  File "/var/task/c7n/policy.py", line 450, in run
    return self.run_resource_set(event, resources)
  File "/var/task/c7n/policy.py", line 480, in run_resource_set
    results = action.process(resources)
  File "/var/task/c7n/resources/securityhub.py", line 409, in process
    finding = self.get_finding(
  File "/var/task/c7n/resources/securityhub.py", line 531, in get_finding
    finding_resources.append(self.format_resource(r))
  File "/var/task/c7n/resources/ebs.py", line 565, in format_resource
    details['CreateTime'] = details['CreateTime'].isoformat()

--- a/c7n/query.py
+++ b/c7n/query.py
@@ -342,7 +342,7 @@ class ConfigSource:
             item_config = json.loads(item['configuration'])
         else:
             item_config = item['configuration']
-        return camelResource(item_config)
+        return camelResource(item_config, implicitDate=True)
 
     def get_listed_resources(self, client):
         # fallback for when config decides to arbitrarily break select

--- a/c7n/resources/ec2.py
+++ b/c7n/resources/ec2.py
@@ -1149,7 +1149,8 @@ class InstanceFinding(PostFinding):
             details["VpcId"] = r["VpcId"]
         if "SubnetId" in r:
             details["SubnetId"] = r["SubnetId"]
-        if "IamInstanceProfile" in r:
+        # config will use an empty key
+        if "IamInstanceProfile" in r and r['IamInstanceProfile']:
             details["IamInstanceProfileArn"] = r["IamInstanceProfile"]["Arn"]
 
         instance = {

--- a/c7n/resources/s3.py
+++ b/c7n/resources/s3.py
@@ -42,7 +42,6 @@ from botocore.exceptions import ClientError
 
 from collections import defaultdict
 from concurrent.futures import as_completed
-from dateutil.parser import parse as parse_date
 
 try:
     from urllib3.exceptions import SSLError

--- a/c7n/resources/s3.py
+++ b/c7n/resources/s3.py
@@ -110,7 +110,6 @@ class ConfigS3(query.ConfigSource):
 
         # owner is under acl per describe
         resource.pop('Owner', None)
-        resource['CreationDate'] = parse_date(resource['CreationDate'])
 
         for k, null_value in S3_CONFIG_SUPPLEMENT_NULL_MAP.items():
             if cfg.get(k) == null_value:

--- a/c7n/utils.py
+++ b/c7n/utils.py
@@ -16,6 +16,9 @@ import time
 from urllib import parse as urlparse
 from urllib.request import getproxies
 
+
+from dateutil.parser import ParserError, parse as parse_date
+
 from c7n import config
 from c7n.exceptions import ClientError, PolicyValidationError
 
@@ -188,16 +191,34 @@ def chunks(iterable, size=50):
         yield batch
 
 
-def camelResource(obj):
+def camelResource(obj, implicitDate=False):
     """Some sources from apis return lowerCased where as describe calls
 
     always return TitleCase, this function turns the former to the later
+
+    implicitDate ~ automatically sniff keys that look like isoformat date strings
+     and convert to python datetime objects.
     """
     if not isinstance(obj, dict):
         return obj
     for k in list(obj.keys()):
         v = obj.pop(k)
         obj["%s%s" % (k[0].upper(), k[1:])] = v
+        if implicitDate:
+            # config service handles datetime differently then describe sdks
+            # the sdks use knowledge of the shape to support language native
+            # date times, while config just turns everything into a serialized
+            # json with mangled keys without type info. to normalize to describe
+            # we implicitly sniff keys which look like datetimes, and have an
+            # isoformat marker ('T').
+            kn = k.lower()
+            if isinstance(v, str) and ('time' in kn or 'date' in kn) and "T" in v:
+                try:
+                    dv = parse_date(v)
+                except ParserError:
+                    pass
+                else:
+                    obj["%s%s" % (k[0].upper(), k[1:])] = dv
         if isinstance(v, dict):
             camelResource(v)
         elif isinstance(v, list):

